### ISSUE TITLE
Support New Resource and Data Source for EC2 Managed Prefix Lists

### ISSUE THREAD

**Issue by ewbankkit** [hashicorp/terraform-provider-aws#13986]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> Virtual Private Cloud (VPC) customers can now create and manage their own Prefix Lists to simplify VPC configurations.

##### New

* `aws_ec2_managed_prefix_list`

##### Affected

* `aws_route_table`
* `aws_route`

### Potential Terraform Configuration

```hcl
resource "aws_ec2_managed_prefix_list" "example" {
  address_family = "IPv4"
  max_entries    = 1
  
  entry {
    cidr        = "0.0.0.0/0"
    description = "Everywhere"
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/06/amazon-virtual-private-cloud-customers-use-prefix-lists-simplify-configuration-security-groups-route-tables/).
[User guide](https://docs.aws.amazon.com/vpc/latest/userguide/managed-prefix-lists.html).

Requires [AWS SDK v1.32.12](https://github.com/aws/aws-sdk-go/releases/tag/v1.32.12):
* https://github.com/terraform-providers/terraform-provider-aws/pull/13803

**Comment by roberth-k** [comment]
I'll work on a PR for this.

Additional affected resources:

`aws_security_group`
`aws_security_group_rule`

**Comment by ewbankkit** [comment]
@roberth-k Thanks for offering to tackle this.
I am working on some other fixes/enhancements for routes and route tables and can do the work to change `destination_prefix_list_id` from a `Computed`-only attribute to an `Optional` and `Computed` attribute.
Let us know if you need any assistance.

**Comment by roberth-k** [comment]
Thanks @ewbankkit -- if you could update `destination_prefix_list_id` in `aws_route` it would be helpful. This attribute should be added to the matching data resource as well. One of the vendor prefix lists such as `com.amazonaws.eu-west-1.s3` (via `data_source_aws_prefix_list`) should work for acceptance testing.

**Comment by yermulnik** [comment]
Any idea re approx ETA for this feature release?

**Comment by forty** [comment]
Just a note that support could also be added to transit gateway route table objects https://aws.amazon.com/fr/about-aws/whats-new/2020/08/amazon-transit-gateway-customers-use-own-prefix-lists-simplify-ip-management/

**Comment by roshpr** [comment]
Any idea on the release this feature will be available?

**Comment by rryke** [comment]
also looking forward to this feature

**Comment by ewbankkit** [comment]
[Blog post](https://aws.amazon.com/blogs/networking-and-content-delivery/simplify-network-routing-and-security-administration-with-vpc-prefix-lists/).

**Comment by Hany-Benz** [comment]
Hello here, 

Any update regarding the release of this feature ?

**Comment by scott881994** [comment]
Will this feature also add an aws_managed_prefix_list data source as well?

**Comment by Tensho** [comment]
@scott881994 https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/prefix_list

**Comment by JosephHobbs** [comment]
Just for clarification, there are essentially two types of prefix lists in AWS: AWS managed and Customer Managed.  The existing prefix_list data source you reference uses the DescribePrefixLists API (https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribePrefixLists.html) and will only return AWS managed prefix lists.

What I believe we're looking for is a data_source to reference Customer managed prefix lists which would need to call the DescribeManagedPrefixLists API (https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeManagedPrefixLists.html).  To take it one step further, I'd love to see the ability to also create/manage customer-managed prefix lists.

**Comment by roberth-k** [comment]
Hi @JosephHobbs; https://github.com/terraform-providers/terraform-provider-aws/pull/14110 updates the `aws_prefix_list` data source such that it returns both AWS-managed and customer-managed prefix lists. The PR goes into detail about why, but in essence, the managed prefix list API-s are a superset, and the same is reflected in CloudFormation as well.

**Comment by JosephHobbs** [comment]
> Hi @JosephHobbs; #14110 updates the `aws_prefix_list` data source such that it returns both AWS-managed and customer-managed prefix lists. The PR goes into detail about why, but in essence, the managed prefix list API-s are a superset, and the same is reflected in CloudFormation as well.

That's awesome.  Thanks for the clarification/information!  I can't wait until this feature is available...

**Comment by breathingdust** [comment]
Hi all! :wave: Just wanted to direct you to our public [roadmap](https://github.com/hashicorp/terraform-provider-aws/blob/master/ROADMAP.md) for this quarter (Nov-Jan) in which this item has been mentioned.

Due to the significant community interest in support for this feature, we will be looking at merging existing contributions soon. 

We appreciate all the contributions and feedback thus far.

Look out for support in the provider soon!

**Comment by bflad** [comment]
To provide an update, this is now under active review and should be released within the next 3 weeks or so. Shout out to @roberth-k for the implementation efforts so far. 🎉

**Comment by bflad** [comment]
For housekeeping purposes, please note that this issue will be specifically for tracking the creation of the following (and I will update the title):

* New Data Source: `aws_ec2_managed_prefix_list`
* New Resource: `aws_ec2_managed_prefix_list`

For following EC2 Route and EC2 Route Table updates, please see: https://github.com/hashicorp/terraform-provider-aws/issues/15273

For following EC2 Transit Gateway Prefix List Reference updates, please see: https://github.com/hashicorp/terraform-provider-aws/issues/16572

If there is other functionality that should be updated as part of supporting EC2 Managed Prefix Lists, please create other new GitHub issues. 👍

**Comment by ahmgithubahm** [comment]
> If there is other functionality that should be updated as part of supporting EC2 Managed Prefix Lists, please create other new GitHub issues. 👍

Just to clarify, if we wanted support for using user-managed prefix lists in ingress and egresss security group rules, that'd be under a new issue?

cheers

**Comment by bflad** [comment]
@ahmgithubahm great question, and yes please, create a separate issue to ensure we do not miss that!

**Comment by roberth-k** [comment]
@ahmgithubahm 

> Just to clarify, if we wanted support for using user-managed prefix lists in ingress and egresss security group rules, that'd be under a new issue?
> 
> cheers

AWS-managed and (customer) Managed Prefix Lists form a shared namespace, i.e. the `prefix_list_ids` of a security group rule should support both.

**Comment by ahmgithubahm** [comment]
I was half-way through raising the enhancement request, but thought I'd better check what happens if you try using the prefix_list_ids argument on an ingress rule in an aws_security_group_rule - and ... it just works.  Obviously I hard-coded the pl-nnnn id, but the ingress rule was successfully created.

It seems that despite the documentation for prefix_list_ids in aws_security_group_rule saying it's for egress only, this is not actually enforced by Terraform (AWS provider 3.20.0), so can be successfully used for ingress as well, with any valid prefix list ids, even customer-managed.

I have not tested this in the aws_security_group resource, since I set all security group rules separately.

**Comment by bflad** [comment]
Support for the new `aws_ec2_managed_prefix_list` resource and data source has been merged in and will release with version 3.22.0 of the Terraform AWS Provider, likely tomorrow. Thank you to @roberth-k for the implementation work. 👍 

For operators in AWS GovCloud (US) and potentially other partitions, please note that the EC2 API does not yet support this functionality.

Followup issues for tracking:

- EC2 Route and EC2 Route Table updates, please see: #15273
- EC2 Transit Gateway Prefix List Reference updates, please see: #16572
